velox/dwio/common/CachedBufferedInput.cpp

```diff
@@ -248,12 +255,21 @@ void CachedBufferedInput::makeLoads(
         readRegion(ranges, prefetch);
       });
   if (prefetch && executor_) {
-    for (auto& load : allCoalescedLoads_) {
+    std::vector<int32_t> doneIndices;
+    for (auto i = 0; i < allCoalescedLoads_.size(); ++i) {
+      auto& load = allCoalescedLoads_[i];
       if (load->state() == LoadState::kPlanned) {
         executor_->add([pendingLoad = load]() {
           process::TraceContext trace("Read Ahead");
           pendingLoad->loadOrFuture(nullptr);
         });
+      } else {
+        doneIndices.push_back(i);
+      }
+      // Remove the loads that were done. There can be done loads if the same
+      // CachedBufferedInput has multiple cycles of enqueues and loads.
+      for (int32_t i = doneIndices.size() - 1; i >= 0; --i) {
+        allCoalescedLoads_.erase(allCoalescedLoads_.begin() + doneIndices[i]);
```

Comment:
This should be outside the looping over `allCoalescedLoads_`. 

Also this is more efficient if `allCoalescedLoads_` is large:

```c++
std::swap(allCoalescedLoads_[doneIndices[i]], allCoalescedLoads_.back());
allCoalescedLoads_.pop_back();
```